File: src/commands/patterns.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::auth;

#[derive(Deserialize, Serialize)]
struct PatternItem {
    id: String,
    title: String,
    description: String,
    frequency: String,
    significance: String,
    raw_sequence: serde_json::Value,
    generated_at: String,
}

#[derive(Deserialize, Serialize)]
struct EnrichmentResponse {
    enrichment_status: String,
    data: Vec<PatternItem>,
}

/// View behavioral patterns for a product.
pub async fn run(product_id: &str, api_base: &str, json: bool) -> Result<()> {
    let token = auth::get_valid_access_token().await?;
    let client = reqwest::Client::new();
    let url = format!("{api_base}/products/{product_id}/patterns");

    let response = client
        .get(&url)
        .bearer_auth(&token)
        .send()
        .await
        .context("failed to reach the Oriyn API")?;

    let status = response.status();
    if !status.is_success() {
        let body = response
            .text()
            .await
            .unwrap_or_else(|_| "(no body)".to_string());
        anyhow::bail!("API returned {status}: {body}");
    }

    let resp: EnrichmentResponse = response
        .json()
        .await
        .context("failed to parse patterns response")?;

    if json {
        println!(
            "{}",
            serde_json::to_string(&resp).context("failed to serialize")?
        );
        return Ok(());
    }

    println!("Enrichment status: {}", resp.enrichment_status);
    println!();

    if resp.data.is_empty() {
        println!("No patterns found.");
        return Ok(());
    }

    for pattern in &resp.data {
        println!("{}", pattern.title);
        println!("  {}", pattern.description);
        println!("  Frequency:    {}", pattern.frequency);
        println!("  Significance: {}", pattern.significance);
        if let Some(steps) = pattern.raw_sequence.as_array() {
            let labels: Vec<&str> = steps.iter().filter_map(|s| s.as_str()).collect();
            if !labels.is_empty() {
                println!("  Sequence:     {}", labels.join(" -> "));
            }
        }
        println!();
    }
    Ok(())
}

File: cmd/patterns.go
package cmd

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/spf13/cobra"
)

func newPatternsCmd(app *App) *cobra.Command {
	var productID string
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:   "patterns",
		Short: "View behavioral patterns for a product",
		RunE: func(cmd *cobra.Command, args []string) error {
			resp, err := app.API.GetPatterns(cmd.Context(), productID)
			if err != nil {
				return err
			}
			app.Tracker.Capture("cli_patterns_viewed", map[string]interface{}{"product_id": productID})

			w := cmd.OutOrStdout()
			if jsonOutput {
				return printJSON(w, resp)
			}

			fmt.Fprintf(w, "Enrichment status: %s\n\n", resp.EnrichmentStatus)

			if len(resp.Data) == 0 {
				fmt.Fprintln(w, "No patterns found.")
				return nil
			}

			for _, pattern := range resp.Data {
				fmt.Fprintln(w, pattern.Title)
				fmt.Fprintf(w, "  %s\n", pattern.Description)
				fmt.Fprintf(w, "  Frequency:    %s\n", pattern.Frequency)
				fmt.Fprintf(w, "  Significance: %s\n", pattern.Significance)
				var steps []string
				if err := json.Unmarshal(pattern.RawSequence, &steps); err == nil && len(steps) > 0 {
					fmt.Fprintf(w, "  Sequence:     %s\n", strings.Join(steps, " -> "))
				}
				fmt.Fprintln(w)
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&productID, "product-id", "", "The product ID")
	_ = cmd.MarkFlagRequired("product-id")
	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output raw JSON")
	return cmd
}
